### PR TITLE
Batch of Wayland improvements and fixes.

### PR DESCRIPTION
**Pull request by Kontrabant**
A batch of Wayland improvements and fixes.

The big changes are additional refactoring to fullscreen handling in order to properly handle the compositor initiating a fullscreen enter/leave event, or when the compositor moves a fullscreen window between displays.  Previously, SDL's internal state wasn't being updated with these events, so the SDL window state could get out of sync with the actual window state and lead to unexpected behavior (similar to what was reported in #6153 on x11). SDL's internal fullscreen state tracking is now properly updated with externally initiated window state changes, and things like toggling fullscreen or moving fullscreen windows between displays via a desktop hotkey will now behave properly.

Aside from that, there are other small fixes:

- Add some missing `static` qualifiers.
- Commit in the Wayland backend when changing the fullscreen flags or parameters so the changes take immediate effect.
- Work around a [libdecor bug](https://gitlab.gnome.org/jadahl/libdecor/-/issues/40) that could result in a window size change when hiding a floating window.
- Fix a bug that prevented a resize event from being generated when the display scale changed.
- A change to silence some errant log messages in the video core when mode switching is disabled.
- Some other minor code cleanup in the Wayland backend (a function parameter was no longer needed and some common code was moved to a separate function).

### CODE REVIEW

File: src/video/wayland/SDL_waylandwindow.c [hunk 1]
```diff
@@ -448,7 +453,74 @@ SetFullscreen(SDL_Window *window, struct wl_output *output, SDL_bool commit)
...
+                       (!window->fullscreen_mode.w || !window->fullscreen_mode.h)) {
+                wind->fullscreen_flags = SDL_WINDOW_FULLSCREEN_DESKTOP;
+            }
+
+            wind->is_fullscreen = SDL_TRUE;
+
+            wind->in_fullscreen_transition = SDL_TRUE;
+            SDL_SetWindowFullscreen(window, wind->fullscreen_flags);
+            wind->in_fullscreen_transition = SDL_FALSE;
+            SetMinMaxDimensions(window, SDL_FALSE);
```
Comment (cgutman) [inline src/video/wayland/SDL_waylandwindow.c:487]: Why the calls to `SetMinMaxDimensions()` here?

File: src/video/wayland/SDL_waylandwindow.c [hunk 2]
```diff
@@ -831,8 +895,15 @@ decoration_frame_configure(struct libdecor_frame *frame,
         height = window->windowed.h;
         wind->floating_resize_pending = SDL_FALSE;
     } else {
+        /*
+         * XXX: When hiding a floating window, libdecor can send bogus content sizes that
+         *      are +/- the height of the title bar, which distorts the window size.
+         *      Ignore any values from libdecor when hiding a floating window.
```
Comment (cgutman) [inline src/video/wayland/SDL_waylandwindow.c:901]: Is there a libdecor bug for this?

Comment (Kontrabant) [reply src/video/wayland/SDL_waylandwindow.c:901]: It's linked in the pull request.

https://gitlab.gnome.org/jadahl/libdecor/-/issues/40

File: src/video/wayland/SDL_waylandwindow.c (hunk 1 above)
Comment (Kontrabant) [reply src/video/wayland/SDL_waylandwindow.c:487]: In particular, the max dimension _must_ be disabled when entering fullscreen as entering fullscreen doesn't mean that window size limits will be ignored .

Unsetting the min/max when entering fullscreen via a compositor event might not be needed, as most compositors seem to disable switching into fullscreen externally if the window is non-resizeable or has a max dimension smaller than the output, but there's nothing guaranteeing this, and this keeps the end state the same whether fullscreen is toggled internally via an SDL function or externally via the compositor.  The limits always need to be restored on exiting fullscreen.

They are unset/set when entering/exiting fullscreen via the internal SDL functions as well.

**Review by flibitijibibo — COMMENTED** [review]
Changes look okay (minus the parts @cgutman is already asking about), but as always I'm skeptical of changes that involve commits and fullscreen/windowed changes since we constantly get dinged for doing this wrong (i.e. "only commit once and only once for window changes") and always seems to break games in the wild (my usual tests are Super Hexagon and FEZ, SupHex might need reverting to the pre-Neo version though).

As long as this gets thrown at a lot of different applications I'm willing to look the other way since I can't do very much work on this myself right now...

**Comment by Kontrabant** [comment]
The bit about committing when changing fullscreen flags is necessary to propagate changes when changing from SDL_WINDOW_FULLSCREEN ←→ SDL_WINDOW_FULLSCREEN_DESKTOP.  The SDL video core updates the flags, then calls into the backend `SetWindowFullscreen()` function to notify the backend of the change, so we need to respond to it or the changes won't take effect until something else causes a commit, such as a focus change. 

Overall though, this, along with some previous work, actually completely eliminates unnecessary fullscreen state toggles and considerably cuts down on commits regarding fullscreen/windowed state setting. I dare say the Wayland backend is actually quite well behaved in that regard now.

**Comment by slouken** [comment]
Okay, let's go ahead and merge for testing.

**Pull request merged**